### ISSUE TITLE
auto generate id

### ISSUE THREAD

**Issue by ccfiel** [mobxjs/mobx-state-tree#780]
I would like to have a properties that before created I would assign a value like generating a UUID for now I use types.optional to do that and assigning a default value. Is there another way to it?

```
export const Category = types
  .model("Category", {
    id: types.optional(types.identifier(), generateId()),
    name: types.string,
    color: types.string,
  })
  .actions(self => ({
    edit(id, data) {
      self.done = !self.done;
    }
  }));
```

**Comment by luisherranz** [comment]
That's the way we do it as well.

**Comment by robinfehr** [comment]
in the beginning I missed an autogenerated uuid too. 
but there are some reasons i'd prefer not to add such feature for ids.

what might help is:
types.optional as you already use it or a custom type.

the reasons i'd prefer not to have it for ids - oh I guess that above had been enough too but:
not only one might want to use something else than a uuid using the types.string to refine the identifier but also in the uuid itself there are a lot of differences 

there are cases where a uuidv4 (all parts of the uuid random) isn't a good choice since you might wan't them to be order-able. a scenario would be storing user id's on a database. you probably don't want that to be purely random, and therefore scattered because filtering and inserting could be slower over time. what mongodb there does for exmaple is providing the objectid a custom id that helps with that.

on the other hand sometimes a v4 might be a good choice. e.g. when you need the entropy and don't want a v1 where the MAC address could be spoofed. a use case could be session ids...then also you can chose from the other versions for other use cases. 

then there all the other versions which you'd have to evaluate for your case. e.g. you software is distributed...

blabla ...point is...it is quite a use case specific thing.

**Comment by luisherranz** [comment]
> blabla ...point is...it is quite a use case specific thing.

😆 

Now that I think it twice, we moved away from `uuidv4` because we started using `toMatchSnapshot` in our tests. We switched to unique ids, based on the object props at the moment of creation and injected automatically in our `addSomething()` actions.

**Comment by ccfiel** [comment]
I was able to solve this using preProcessSnapshot

```
function assignUUID(snapshot) {
  if (!isUUID(snapshot.id)) {
    return Object.assign({}, snapshot, { id: generateId() });
  } else {
    return Object.assign({}, snapshot);
  }
}

export const Category = types
  .model("Category", {
    id: types.identifier(),
    name: types.string,
    color: types.optional(types.string, ""),
  })
  .preProcessSnapshot(snapshot => assignUUID(snapshot))
  .actions(self => ({
    edit(data) {
      Object.keys(data).forEach(function(key) {
        if (!(key === "id")) {
          self[key] = data[key];
        }
      });
    },
    delete() {
      getRoot(self).delete(self);
    },
  }));

```

**Comment by robinfehr** [comment]
ah sure that's another way to go. i guess we're good to close that.

**Comment by ccfiel** [comment]
why this is reopened? :)

**Comment by robinfehr** [comment]
was wondering if a fn could be provided to the types.identifier to provide a custom identifier :)

**Comment by ccfiel** [comment]
that would be a great idea :smiley:

**Comment by robinfehr** [comment]
i guess code wise it is be doable unless i miss something 🥜

when a second param (the id generator fn) would be present within the types.identifer fn we could instead of just returning the IdentifierType wrap such with a types.optional like it's done for the types.maybe and add the id gen fn as the default. 

so it would basically be a syntax sugar for the types.optional version.

@mweststrate what do you think about this syntax sugar? 
i guess it would be quite intuitive.

**Comment by mweststrate** [comment]
Would this be solved by #887 as well?

**Comment by mweststrate** [comment]
I think offering many syntactic variations to write the same type gets confusing, I propose to not do this for now.
